File: Products/sample-android/app/src/main/java/mockifer/example/android/main/MainActivity.java
package mockifer.example.android.main;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.gson.Gson;
import com.squareup.picasso.Picasso;

import java.io.IOException;

import mockifer.Mockifer;
import mockifer.example.android.R;
import mockifer.example.android.cats.Cat;
import mockifer.example.android.cats.Cats;
import mockifer.example.android.common.Requester;
import mockifer.example.android.edit.EditCatActivity;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Response;

import static android.support.v7.widget.LinearLayoutManager.VERTICAL;

public class MainActivity extends AppCompatActivity {
    Button resetButton;
    Button mockErrorButton;
    RecyclerView recyclerView;

    Cat[] cats = new Cat[0];

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        resetButton = (Button) findViewById(R.id.resetButton);
        mockErrorButton= (Button) findViewById(R.id.mockErrorButton);

        recyclerView = (RecyclerView) findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, VERTICAL, false));
        recyclerView.setAdapter(new CatsAdapter());

        resetButton.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View view) {
                resetServerData();
            }
        });

        mockErrorButton.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View view) {
                // Push a mock into the collection of routes that are evaluated for requests.
                Mockifer.pushMock("mocks.cats.getcats.error");
                loadCats();
            }
        });
    }

    @Override protected void onResume() {
        super.onResume();
        loadCats();
    }

    void resetServerData() {
        Mockifer.reset();
        loadCats();
    }

    void loadCats() {
        Requester.doRequest("GET", "/cats", null, new Callback() {
            @Override public void onFailure(@NonNull Call call, @NonNull IOException e) {
            }

            @Override public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                final int statusCode = response.code();
                final String responseBody = response.body().string();

                if (!response.isSuccessful()) {
                    runOnUiThread(new Runnable() {
                        @Override public void run() {
                            showErrorDialog("Status code: " + statusCode + "\n\n" + responseBody);
                        }
                    });
                    return;
                }

                cats = new Gson().fromJson(responseBody, Cats.class).cats;
                runOnUiThread(new Runnable() {
                    @Override public void run() {
                        recyclerView.getAdapter().notifyDataSetChanged();
                    }
                });
            }
        });
    }

    private void showErrorDialog(String message) {
        new AlertDialog.Builder(this)
                .setTitle("Network Error")
                .setMessage(message)
                .setPositiveButton("OK", null)
                .create()
                .show();
    }

    private void catSelected(int position) {
        Intent intent = new Intent(this, EditCatActivity.class);
        intent.putExtra(EditCatActivity.EXTRA_CAT, cats[position]);
        startActivity(intent);
    }

    private class CatsAdapter extends RecyclerView.Adapter<CatsAdapter.ViewHolder> {
        @Override public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.cat_list_item, parent, false));
        }

        @Override public void onBindViewHolder(ViewHolder holder, int position) {
            holder.bind(cats[position]);
        }

        @Override public int getItemCount() {
            return cats.length;
        }

        class ViewHolder extends RecyclerView.ViewHolder {
            private ImageView catImageView;
            private TextView catIdTextView;
            private TextView catNameTextView;
            private TextView catAgeTextView;

            ViewHolder(View itemView) {
                super(itemView);
                catImageView = (ImageView) itemView.findViewById(R.id.catImage);
                catIdTextView = (TextView) itemView.findViewById(R.id.catId);
                catNameTextView = (TextView) itemView.findViewById(R.id.catName);
                catAgeTextView = (TextView) itemView.findViewById(R.id.catAge);

                itemView.setOnClickListener(new View.OnClickListener() {
                    @Override public void onClick(View view) {
                        catSelected(getAdapterPosition());
                    }
                });
            }

            void bind(Cat cat) {
                Picasso.with(itemView.getContext()).load(Requester.BASE_URL + cat.image).into(catImageView);
                catIdTextView.setText("Id:     " + cat.id);
                catNameTextView.setText("Name: " + cat.name);
                catAgeTextView.setText("Age:   " + cat.age);
            }
        }
    }
}

File: support/samples/sample-android/app/src/main/java/io/github/marcelbraghetto/mockifer/example/android/main/MainActivity.kt
package io.github.marcelbraghetto.mockifer.example.android.main

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager.VERTICAL
import androidx.recyclerview.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.google.gson.Gson
import com.squareup.picasso.Picasso
import io.github.marcelbraghetto.mockifer.example.android.BuildConfig
import io.github.marcelbraghetto.mockifer.example.android.R
import io.github.marcelbraghetto.mockifer.example.android.cats.Cat
import io.github.marcelbraghetto.mockifer.example.android.cats.Cats
import io.github.marcelbraghetto.mockifer.example.android.common.Requester
import kotlinx.android.synthetic.main.activity_main.*
import io.github.marcelbraghetto.mockifer.example.android.edit.EditCatActivity
import io.github.marcelbraghetto.mockifer.example.android.MockiferUtil
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response
import java.io.IOException

class MainActivity : AppCompatActivity() {
    private var cats = arrayOf<Cat>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        recyclerView.run {
            layoutManager = LinearLayoutManager(this@MainActivity, VERTICAL, false)
            adapter = CatsAdapter()
        }

        if (BuildConfig.IS_MOCK_APP) {
            resetButton.setOnClickListener { resetServerData() }

            mockErrorButton.setOnClickListener {
                // Push a mock into the collection of routes that are evaluated for requests.
                MockiferUtil.pushMock(mockId = "mocks.cats.getcats.error")
                loadCats()
            }
        } else {
            resetButton.visibility = GONE
            mockErrorButton.visibility = GONE
        }
    }

    override fun onResume() {
        super.onResume()
        loadCats()
    }

    private fun resetServerData() {
        MockiferUtil.reset()
        loadCats()
    }

    private fun loadCats() {
        Requester.doRequest("GET", "/cats", null, object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                Log.d("BLAH", "Load cats failed: ${e.localizedMessage}")
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                val statusCode = response.code()
                val responseBody = response.body()!!.string()

                if (!response.isSuccessful) {
                    runOnUiThread { showErrorDialog("Status code: " + statusCode + "\n\n" + responseBody) }
                    return
                }

                cats = Gson().fromJson(responseBody, Cats::class.java).cats
                runOnUiThread { recyclerView.adapter?.notifyDataSetChanged() }
            }
        })
    }

    private fun showErrorDialog(message: String) {
        AlertDialog.Builder(this)
                .setTitle("Network Error")
                .setMessage(message)
                .setPositiveButton("OK", null)
                .create()
                .show()
    }

    private fun catSelected(position: Int) {
        startActivity(Intent(this, EditCatActivity::class.java).apply {
            putExtra(EditCatActivity.EXTRA_CAT, cats[position])
        })
    }

    private inner class CatsAdapter : RecyclerView.Adapter<CatsAdapter.ViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.cat_list_item, parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.bind(cats[position])
        }

        override fun getItemCount(): Int {
            return cats.size
        }

        internal inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            private val imageView = itemView.findViewById<ImageView>(R.id.catImageView)
            private val idTextView = itemView.findViewById<TextView>(R.id.catIdTextView)
            private val nameTextView = itemView.findViewById<TextView>(R.id.catNameTextView)
            private val ageTextView = itemView.findViewById<TextView>(R.id.catAgeTextView)

            init {
                itemView.setOnClickListener { catSelected(adapterPosition) }
            }

            fun bind(cat: Cat) {
                Picasso.with(itemView.context)
                        .load(BuildConfig.SERVER_BASE_URL + cat.image)
                        .into(imageView)

                idTextView.text = "Id: ${cat.id}"
                nameTextView.text = "Name: ${cat.name}"
                ageTextView.text = "Age: ${cat.age}"
            }
        }
    }
}
